components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -5,68 +5,565 @@
+  if (it == signals.end()) {
+    return nullptr;
+  }
+  return &it->second;
+}
+
+double GetPopRecency(const mojom::FeedItemMetadataPtr& article) {
+  auto& publish_time = article->publish_time;
+
+  // TODO(fallaciousreasoning): Use the new popularity field instead.
+  // https://github.com/brave/brave-browser/issues/32173
+  double popularity = article->score == 0 ? 50 : article->score;
+  double multiplier = publish_time > base::Time::Now() - base::Hours(5) ? 2 : 1;
+  auto dt = base::Time::Now() - publish_time;
+
+  return multiplier * popularity *
+         pow(0.5, dt.InHours() / kPopRecencyHalfLifeInHours);
+}
+
+double GetArticleWeight(const mojom::FeedItemMetadataPtr& article,
+                        const Signal& signal) {
+  double source_visits_projected =
+      kSourceVisitsMin + signal.visit_weight * (1 - kSourceVisitsMin);
+  double source_subscribed_projected =
+      signal.subscribed ? kSourceSubscribedMax : kSourceSubscribedMin;
+  return source_visits_projected * source_subscribed_projected *
+         GetPopRecency(article);
+}
+
+std::string PickRandom(const std::vector<std::string>& items) {
+  CHECK(!items.empty());
+  // Note: RandInt is inclusive, hence the minus 1
+  return items[base::RandInt(0, items.size() - 1)];
+}
+
+ArticleInfos GetArticleInfos(const FeedItems& feed_items,
+                             const Signals& signals) {
+  ArticleInfos articles;
+  base::flat_set<GURL> seen_articles;
+  for (const auto& item : feed_items) {
+    if (item.is_null()) {
+      continue;
+    }
+    if (item->is_article()) {
+      // Because we download feeds from multiple locales, it's possible there
+      // will be duplicate articles, which we should filter out.
+      if (seen_articles.contains(item->get_article()->data->url)) {
+        continue;
+      }
+      auto& article = item->get_article();
+
+      seen_articles.insert(article->data->url);
+
+      const auto* signal = GetSignal(article->data, signals);
+
+      // If we don't have a signal for this article, filter it out.
+      if (!signal) {
+        continue;
+      }
+
+      auto pair = std::tuple(article->data->Clone(), *signal,
+                             GetArticleWeight(article->data, *signal));
+      articles.push_back(std::move(pair));
+    }
+  }
+  return articles;
+}
+
+// Randomly true/false with equal probability.
+bool TossCoin() {
+  return base::RandDouble() < 0.5;
+}
+
+// This is a Box Muller transform for getting a normally distributed value
+// between [0, 1] (inclusive)
+// https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform
+double GetNormal() {
+  double u;
+  double v;
+
+  do {
+    u = base::RandDouble();
+  } while (u == 0);
+  do {
+    v = base::RandDouble();
+  } while (v == 0);
+
+  double result = sqrt(-2.0 * log(u)) * cos(2 * 3.1415 * v);
+  result = result / 10 + 0.5;
+
+  // Resample if outside the [0, 1] range
+  if (result > 1 || result < 0) {
+    return GetNormal();
+  }
```

Comment:
I'm not that good at math 😭. Having that said, I'm curious if the probability of picking values outside ranges is less than the case we want? If not, Could interpolating the sampled value be an alternative for resampling?

Reply:
So I think the probability of being outside the 0-1 range is about 2%, so the likelihood of it happening a few times in a row is pretty small (the likelihood of 5 resamples is 3 per billion). The problem with interpolating is that theoretically, any value is possible with a normal distribution, it just gets really unlikely for a large number of standard deviations.

Reply:
Great!